### PR TITLE
Add api for the percentage of unsigned batches monitoring

### PR DESCRIPTION
**Pull request by wmb-software-consulting**
## Why are these changes needed?

Add an API to dataapi server for the percentage of unsigned batches monitoring 

## Checks

- [x] I've made sure the lint is passing in this PR.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, in that case, please comment that they are not relevant.
- Testing Strategy
   - [x] Unit tests
   - [ ] Integration tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: disperser/dataapi/server.go
Comment (jianoaix) [inline disperser/dataapi/server.go:72]: We actually need this:
- `TotalUnsignedBatches` per operator: as the name indicates, we need to know how many unsigned batches each operator has (we interest in those who have greater than 0 unsigned batches)
- `TotalBatches`: this is what you already defined

With the above two, what we want to get is that for each operator, what's the percentage of their unsigned batches. This is `percent[operator] = TotalUnsignedBatches[operator] / TotalBatches`.

**Comment by jianoaix** [comment]
Can you merge master to sync the PR to the latest?

File: disperser/dataapi/docs/swagger.yaml
Comment (jianoaix) [inline disperser/dataapi/docs/swagger.yaml:307]: Here and other relevant locations: can we name it as `/metrics/operator_nonsigning_percentage`? 
In this API, we are interested in learning what's the nonsigning percentage for each operator.

File: disperser/dataapi/server.go
Comment (jianoaix) [inline disperser/dataapi/server.go:378]: Similar to name this handler properly

File: disperser/dataapi/metrics_handlers.go
Comment (jianoaix) [inline disperser/dataapi/metrics_handlers.go:165]: @gpsanant Is this the right way to get total number of batches in the past N seconds?

File: disperser/dataapi/metrics_handlers.go
Comment (jianoaix) [inline disperser/dataapi/metrics_handlers.go:180]: If `totalOfBatches` is 0, we can return immediately, right?

Comment (wmb-software-consulting) [reply disperser/dataapi/metrics_handlers.go:180]: Actually, we need to return `map[string]float64` and  `nonSigners` is type of `map[string]int`. 
If we return right away we don't have the operators' id inside map

File: disperser/dataapi/docs/swagger.json
Comment (jianoaix) [inline disperser/dataapi/docs/swagger.json:475]: NonsigningPercentage?

File: disperser/dataapi/server.go
Comment (jianoaix) [inline disperser/dataapi/server.go:72]: Can we rename `UnsignedBatches` to `OperatorNonsigningPercentage`? (and fix other places accordingly)

File: disperser/dataapi/metrics_handlers.go
Comment (jianoaix) [inline disperser/dataapi/metrics_handlers.go:156]: nit: just `totalBatches`

File: disperser/dataapi/metrics_handlers.go
Comment (jianoaix) [reply disperser/dataapi/metrics_handlers.go:180]: If `totalBatches` is 0, then all operators will have 0% nonsigning batches, right?

Also for the `map[string]float64`, we only need to return operators who have >0% nonsigning percentage.

**Review by jianoaix — COMMENTED** [review]
Almost there!

**Review by jianoaix — APPROVED** [review]
LGTM. Maybe @gpsanant can have a quick check, but otherwise LGTM.

File: disperser/dataapi/metrics_handlers.go
Comment (jianoaix) [inline disperser/dataapi/metrics_handlers.go:194]: Can we just use `totalUnsignedBatches > 0`? It's integer, so play better with comparison.

**Review by jianoaix — APPROVED** [review]

File: disperser/dataapi/docs/swagger.yaml
Comment (jianoaix) [inline disperser/dataapi/docs/swagger.yaml:277]: nit: this seems to be from the /metrics/non_signers. We may revise it to fit this API better, like "Interval to query for operator nonsigning percentage"

File: disperser/dataapi/metrics_handlers.go
Comment (jianoaix) [inline disperser/dataapi/metrics_handlers.go:147]: nit: update this function name

File: disperser/dataapi/server.go
Comment (jianoaix) [inline disperser/dataapi/server.go:372]: nit: same as the comment in Swagger, this interval param needs a bit revision

File: disperser/dataapi/metrics_handlers.go
Comment (gpsanant) [reply disperser/dataapi/metrics_handlers.go:165]: nope, we should see if graphql has a better way to get the count of batches or add a running count to the subgraph

File: disperser/dataapi/metrics_handlers.go
Comment (gpsanant) [inline disperser/dataapi/metrics_handlers.go:190]: what if there were a bunch of batches that they weren't an operator during? this number will be much higher than real. we need to find out how many batches they were active for that they signed. i recommend we refactor the subgraph if this is what we want

Comment (gpsanant) [reply disperser/dataapi/metrics_handlers.go:190]: to be more precise, the ratio calculated here is (total batches nonsigned by operator/total batches on eigenda ever), what we want is (total batches nonsigned by operator/total batches on eigenda that operator was requested to sign ever)

File: disperser/dataapi/metrics_handlers.go
Comment (jianoaix) [reply disperser/dataapi/metrics_handlers.go:165]: Having a running count indexed is certainly nice, but with the current subgraph, is there any better way to query it? If not, we should proceed with this way here. We can revisit later if subgraph has that number indexed.

File: disperser/dataapi/metrics_handlers.go
Comment (jianoaix) [reply disperser/dataapi/metrics_handlers.go:190]: Ok, that's good point. Then this number here will be lower than real, right? i.e. the numerator (total unsigned) is accurate, the denominator (the total batches) may be off, if the operator wasn't in network in last N seconds.

File: disperser/dataapi/metrics_handlers.go
Comment (gpsanant) [reply disperser/dataapi/metrics_handlers.go:165]: potentially we could look at the registeration/deregistration events here: https://github.com/wmb-software-consulting/eigenda/blob/12b3081fe87edc0d6e8954f63b00db96e2571c20/subgraphs/eigenda-operator-state/schema.graphql#L12 and query batches within, but its pretty complex

Comment (jianoaix) [reply disperser/dataapi/metrics_handlers.go:165]: It seems doable: given N, we want to find total batches that an operator is responsible for
1) Find all register events in past N secs
2) Find all deregister events in past N secs
3) Using the above two, create a list of disjoint time intervals, where the operator is live in the network
4) Query total number batches in each interval, `Ti = Query(Si, Ei)`
5) The sum of all Ti is the answer

In term of implementation:
The step (3) and (5) are easy.
The step (1) and (2) are about querying the two tables, `OperatorDeregistered` and `OperatorRegistered` that already indexed in subgraph. (This is the least unknow part for me, please correct if I'm wrong)
The step (4) is what's already implemented here.

So overall, I think the question is to fetch all register/deregister events for an operator in past N seconds (step 1 and 2).

Comment (gpsanant) [reply disperser/dataapi/metrics_handlers.go:165]: Yep that's straightforward too. It's just complex to put all these together, but if we have bandwidth, let's do it

**Review by jianoaix — APPROVED** [review]

**Review by jianoaix — APPROVED** [review]
Thanks @wmb-software-consulting for addressing comments.

There may be some other small comments but the PR looks good overall. We will merge this (unless there is any main issue). I'll make follow up fixes if needed.

File: disperser/dataapi/subgraph_client.go
Comment (gpsanant) [inline disperser/dataapi/subgraph_client.go:166]: prob best to set outside loop but ok

File: disperser/dataapi/subgraph_client.go
Comment (gpsanant) [inline disperser/dataapi/subgraph_client.go:168]: you need to filter by reference block number, not timestamp of the batch confirmation

File: disperser/dataapi/subgraph_client.go
Comment (gpsanant) [inline disperser/dataapi/subgraph_client.go:286]: what if they deregister and register in the same block?

Comment (jianoaix) [reply disperser/dataapi/subgraph_client.go:286]: That should not have effect on operator's responsibility: if this block is taken as reference block, then the operator is considered as in the network for signing.

File: disperser/dataapi/subgraph_client.go
Comment (jianoaix) [reply disperser/dataapi/subgraph_client.go:168]: If the block time is monotonically increasing w.r.t. the block number, it should produce equivalent results?
i.e. for block A and B, blockTime(A) < blockTime(B) iff blockNumber(A) < blockNumber(B); if this is true, querying the interval by block number will be equivalent to querying the interval by block time.

File: disperser/dataapi/subgraph_client.go
Comment (jianoaix) [reply disperser/dataapi/subgraph_client.go]: Done

**Comment by jianoaix** [comment]
Merging now, if there is another issue, will follow up

**Pull request merged**
